File: static/casting/sources/byte_model_progressive.c
/* byte_model_progressive.c — progressive architecture search.
 *
 * Architecture family: n → h → 1 MLP with ±1 weights, biases on every
 * unit, sign() activation. For h > 0 total bits = h*(n+2) + 1; the
 * h=0 baseline (linear classifier) is (n+1) bits.
 *
 * Search rule: for each target, grow h = 0, 1, 2, ... until exhaustive
 * search over all 2^W weight bitstrings finds a solver. Record the
 * smallest h that works, and the solver bits. Stop at W > 22 (~4M).
 *
 * This is the capability-per-parameter curve of tiny models. Linear
 * targets (OR/AND/MAJ) solve at h=0 or h=1. XOR needs h=2. Parity
 * over 4 bits may need more. The output is a pool you can serialize
 * and ship; the JS port dumps it as JSON.
 *
 * Compile: cc -O2 -o byte_model_progressive byte_model_progressive.c
 */

#include <stdio.h>
#include <stdint.h>
#include <stdlib.h>
#include <string.h>

#define MAX_H 4
#define MAX_N 4

static int sign_(int x) { return x >= 0 ? +1 : -1; }

/* Forward pass. Weight layout (LSB first):
 *   [input->hidden weights row-major for each hidden unit,
 *    each followed by that hidden unit's bias],
 *   then [output weights for each hidden unit, then output bias].
 * For h==0: just [n input weights, then output bias]. */
static int forward(uint32_t bits, int n, int h, const int *x) {
    int xi[MAX_N];
    for (int i = 0; i < n; ++i) xi[i] = x[i] ? 1 : -1;
    int idx = 0;
    if (h == 0) {
        int s = 0;
        for (int i = 0; i < n; ++i) {
            int w = ((bits >> idx++) & 1) ? 1 : -1;
            s += w * xi[i];
        }
        int b = ((bits >> idx++) & 1) ? 1 : -1;
        return sign_(s + b);
    }
    int hid[MAX_H];
    for (int j = 0; j < h; ++j) {
        int s = 0;
        for (int i = 0; i < n; ++i) {
            int w = ((bits >> idx++) & 1) ? 1 : -1;
            s += w * xi[i];
        }
        int b = ((bits >> idx++) & 1) ? 1 : -1;
        hid[j] = sign_(s + b);
    }
    int s = 0;
    for (int j = 0; j < h; ++j) {
        int w = ((bits >> idx++) & 1) ? 1 : -1;
        s += w * hid[j];
    }
    int b = ((bits >> idx++) & 1) ? 1 : -1;
    return sign_(s + b);
}

static int total_bits(int n, int h) {
    return (h == 0) ? (n + 1) : h * (n + 2) + 1;
}

typedef struct { const char *name; int n; unsigned long tt; } Target;
typedef struct { int solved; int h; int W; uint32_t bits; unsigned long examined; } Solver;

static Solver search_progressive(const Target *tgt, int max_W) {
    Solver s = { 0, 0, 0, 0, 0 };
    int x[MAX_N];
    int N = 1 << tgt->n;
    for (int h = 0; h <= MAX_H; ++h) {
        int W = total_bits(tgt->n, h);
        if (W > max_W) break;
        uint64_t space = 1ULL << W;
        for (uint64_t b = 0; b < space; ++b) {
            int ok = 1;
            for (int row = 0; row < N; ++row) {
                for (int i = 0; i < tgt->n; ++i)
                    x[i] = (row >> (tgt->n - 1 - i)) & 1;
                int y_pred = forward((uint32_t)b, tgt->n, h, x);
                int y_true = ((tgt->tt >> row) & 1) ? +1 : -1;
                if (y_pred != y_true) { ok = 0; break; }
            }
            s.examined++;
            if (ok) { s.solved = 1; s.h = h; s.W = W; s.bits = (uint32_t)b; return s; }
        }
    }
    return s;
}

static Target TARGETS[] = {
    {"2-AND",  2, 0x8},
    {"2-OR",   2, 0xe},
    {"2-XOR",  2, 0x6},
    {"3-AND",  3, 0x80},
    {"3-OR",   3, 0xfe},
    {"3-MAJ",  3, 0xe8},
    {"3-MUX",  3, 0xca},
    {"3-XOR",  3, 0x69},
    {"4-OR",   4, 0xfffe},
    {"4-AND",  4, 0x8000},
    {"4-MAJ",  4, 0xe880},
    {"4-thr2", 4, 0xfee8},
    {"4-XOR",  4, 0x6996},
};
#define N_TARGETS (sizeof(TARGETS)/sizeof(TARGETS[0]))

int main(void) {
    printf("progressive architecture search — tiny MLPs (±1 weights, sign activation)\n");
    printf("grow hidden width h until a solver is found; emit the whole pool.\n\n");
    printf("%-8s  %-3s  %-4s  %-10s  %s\n", "target", "h", "bits", "examined", "solver");
    printf("-----------------------------------------------------------------\n");
    int solved_count = 0;
    for (unsigned i = 0; i < N_TARGETS; ++i) {
        Solver r = search_progressive(&TARGETS[i], 22);
        if (r.solved) {
            printf("%-8s  %-3d  %-4d  %-10lu  0x%x\n",
                   TARGETS[i].name, r.h, r.W, r.examined, r.bits);
            solved_count++;
        } else {
            printf("%-8s  -    -     -           UNSOLVED (W > 22)\n",
                   TARGETS[i].name);
        }
    }
    printf("\npool: %d / %zu targets solved.\n", solved_count, N_TARGETS);
    printf("the JS version serializes this pool as JSON for drop-in use.\n");
    return 0;
}

File: static/casting/js/byte_model_progressive.js
/* Casting: byte_model_progressive — progressive architecture search.
 * Mirrors static/casting/sources/byte_model_progressive.c and also emits
 * the discovered pool as JSON + a download link beneath the output.
 */
(function () {
  const MAX_H = 4;
  const MAX_N = 4;
  const BUDGET_W = 22;

  function totalBits(n, h) { return (h === 0) ? (n + 1) : h * (n + 2) + 1; }
  function forward(bits, n, h, x) {
    const xi = new Array(n);
    for (let i = 0; i < n; ++i) xi[i] = x[i] ? 1 : -1;
    let idx = 0;
    if (h === 0) {
      let s = 0;
      for (let i = 0; i < n; ++i) s += (((bits >>> idx++) & 1) ? 1 : -1) * xi[i];
      s += ((bits >>> idx++) & 1) ? 1 : -1;
      return s >= 0 ? +1 : -1;
    }
    const hid = new Array(h);
    for (let j = 0; j < h; ++j) {
      let s = 0;
      for (let i = 0; i < n; ++i) s += (((bits >>> idx++) & 1) ? 1 : -1) * xi[i];
      s += ((bits >>> idx++) & 1) ? 1 : -1;
      hid[j] = s >= 0 ? +1 : -1;
    }
    let s = 0;
    for (let j = 0; j < h; ++j) s += (((bits >>> idx++) & 1) ? 1 : -1) * hid[j];
    s += ((bits >>> idx++) & 1) ? 1 : -1;
    return s >= 0 ? +1 : -1;
  }

  function searchProgressive(target, maxW) {
    const N = 1 << target.n;
    let examined = 0;
    for (let h = 0; h <= MAX_H; ++h) {
      const W = totalBits(target.n, h);
      if (W > maxW) break;
      const space = 1 << W;
      for (let b = 0; b < space; ++b) {
        examined++;
        let ok = true;
        for (let row = 0; row < N; ++row) {
          const x = new Array(target.n);
          for (let i = 0; i < target.n; ++i) x[i] = (row >> (target.n - 1 - i)) & 1;
          const y_pred = forward(b, target.n, h, x);
          const y_true = ((target.tt >> row) & 1) ? +1 : -1;
          if (y_pred !== y_true) { ok = false; break; }
        }
        if (ok) return { solved: true, h, W, bits: b, examined };
      }
    }
    return { solved: false, h: -1, W: -1, bits: 0, examined };
  }

  const TARGETS = [
    { name: '2-AND',  n: 2, tt: 0x8 },
    { name: '2-OR',   n: 2, tt: 0xe },
    { name: '2-XOR',  n: 2, tt: 0x6 },
    { name: '3-AND',  n: 3, tt: 0x80 },
    { name: '3-OR',   n: 3, tt: 0xfe },
    { name: '3-MAJ',  n: 3, tt: 0xe8 },
    { name: '3-MUX',  n: 3, tt: 0xca },
    { name: '3-XOR',  n: 3, tt: 0x69 },
    { name: '4-OR',   n: 4, tt: 0xfffe },
    { name: '4-AND',  n: 4, tt: 0x8000 },
    { name: '4-MAJ',  n: 4, tt: 0xe880 },
    { name: '4-thr2', n: 4, tt: 0xfee8 },
    { name: '4-XOR',  n: 4, tt: 0x6996 },
  ];

  /* Pool kept across runs of the one-shot too, so "export" always has
   * the last accumulated search. */
  let POOL = [];

  function poolJSON() {
    return JSON.stringify({
      format: 'casting-pool-v1',
      arch_family: 'MLP n→h→1, ±1 weights, sign activation, LSB-first bitstring',
      note: 'Each entry: n inputs, truth_table (2^n bits), h hidden units, '
          + 'W total weight bits, bits = solver (hex). Weights are unpacked '
          + 'LSB-first: for h>0 each hidden unit emits n input weights then 1 bias; '
          + 'then h output weights then 1 output bias. For h=0: n input weights + 1 bias.',
      generated_at: new Date().toISOString(),
      pool: POOL,
    }, null, 2);
  }

  function mountDownload(jsonText) {
    const out = document.getElementById('cast-output');
    if (!out) return;
    let link = document.getElementById('cast-progressive-dl');
    if (!link) {
      link = document.createElement('a');
      link.id = 'cast-progressive-dl';
      link.className = 'cast-chip';
      link.style.display = 'inline-block';
      link.style.marginTop = '0.8em';
      out.insertAdjacentElement('afterend', link);
    }
    if (link.dataset.blobUrl) URL.revokeObjectURL(link.dataset.blobUrl);
    const blob = new Blob([jsonText], { type: 'application/json' });
    const url = URL.createObjectURL(blob);
    link.dataset.blobUrl = url;
    link.href = url;
    link.download = 'casting-progressive-pool.json';
    link.textContent = 'download pool.json (' + POOL.length + ' solvers, ' + blob.size + ' bytes)';
  }

  function run() {
    POOL = [];
    const rows = [];
    rows.push('progressive architecture search — tiny MLPs (±1, sign activation)');
    rows.push('grow h until a solver is found; entire pool is exported below.');
    rows.push('');
    rows.push(['target', 'h', 'bits', 'examined', 'solver (hex)'].map((s, i) =>
      i === 0 ? s.padEnd(8) : i === 4 ? s : s.padEnd(10)
    ).join(' '));
    rows.push('-'.repeat(58));
    let solved = 0;
    for (const t of TARGETS) {
      const r = searchProgressive(t, BUDGET_W);
      if (r.solved) {
        solved++;
        POOL.push({
          name: t.name, n: t.n, truth_table: t.tt,
          h: r.h, weight_bits: r.W, bits: r.bits,
        });
        rows.push(
          t.name.padEnd(8) + ' ' +
          String(r.h).padEnd(10) + ' ' +
          String(r.W).padEnd(10) + ' ' +
          String(r.examined).padEnd(10) + ' ' +
          '0x' + r.bits.toString(16)
        );
      } else {
        rows.push(
          t.name.padEnd(8) + ' ' +
          '-'.padEnd(10) + ' -'.padEnd(11) + ' -'.padEnd(11) +
          ' UNSOLVED (budget W≤' + BUDGET_W + ')'
        );
      }
    }
    rows.push('');
    rows.push('pool: ' + solved + ' / ' + TARGETS.length + ' targets solved.');
    rows.push('');
    const json = poolJSON();
    rows.push('--- exported pool.json below; click the link beneath to download ---');
    rows.push(json);
    mountDownload(json);
    return rows.join('\n');
  }

  window.Casting_byte_model_progressive = { run, poolJSON: () => poolJSON() };
})();
